driver: Add adc driver support for NXP RT series

kele6ra opened the pull request:
I have board with i.MX RT1052 controller and I need to use adc on it. 
I found that Zephyr has external mcux lib with adc driver and wrote adc driver for 1052. 
Then I founf that all RT series has similar adc and made driver for the whole series.
Unfortunately NXP SOC boards don't have adc and I didn't make board tests.

Driver works like adc_mcux_adc12 driver, but have some differences.

I make first pr to zephyr OS, sorry if I make something wrong =)

Signed-off-by: Andrei Ovchinnikov <ovchinnikov@strim-tech.com>

kele6ra commented:
Last test is failed because of https://github.com/zephyrproject-rtos/zephyr/pull/34085

cvinayak commented:
Please rebase to latest upstream/master so that you dont have 189 files of which many are not related to your PR.

kele6ra commented:
> 
> 
> Please rebase to latest upstream/master so that you dont have 189 files of which many are not related to your PR.

Thank you for help, I have rebases my commit. I made mistake trying to add bluetooth fix to my branch,

cvinayak commented:
> 
> 
> > Please rebase to latest upstream/master so that you dont have 189 files of which many are not related to your PR.
> 
> Thank you for help, I have rebases my commit. I made mistake trying to add bluetooth fix to my branch,

You can remove other reviewers to avoid them getting notification emails. (i removed myself out).

kele6ra commented:
> 
> 
> > > Please rebase to latest upstream/master so that you dont have 189 files of which many are not related to your PR.
> > 
> > 
> > Thank you for help, I have rebases my commit. I made mistake trying to add bluetooth fix to my branch,
> 
> You can remove other reviewers to avoid them getting notification emails. (i removed myself out).

Sorry again, I can't find how to change labels and reviewers. They were set by github-actions and maybe I don't have access to remove them?

MaureenHelm reviewed (changes requested):
Thank you for your contribution!

Will you please add `adc` to `boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml` and check that `tests/drivers/adc/adc_api` works? The first part will cause CI to start building this driver for this board.

You will also need to update the hal_nxp repo to build the underlying mcux driver. Please see https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules for more details.

> Sorry again, I can't find how to change labels and reviewers. They were set by github-actions and maybe I don't have access to remove them?

That's correct. I updated the labels.

kele6ra commented on drivers/adc/adc_mcux_rtadc.c:
```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2019 Vestas Wind Systems A/S
+ *
+ * Based on adc_mcux_rt_adc.c, which is:
```
yes, it is based on adc12 driver, add to next commit

kele6ra commented on soc/arm/nxp_imx/rt/Kconfig.defconfig.series:
```diff
@@ -39,6 +39,10 @@ config PWM_MCUX
 	depends on PWM
 
+config ADC_MCUX_RTADC
+	default y
```
add to next commit

kele6ra commented:
> 
> 
> Thank you for your contribution!
> 
> Will you please add `adc` to `boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml` and check that `tests/drivers/adc/adc_api` works? The first part will cause CI to start building this driver for this board.
I have added adc to yaml file, but should I also add board to test file and adc to the board pinmux?

> 
> You will also need to update the hal_nxp repo to build the underlying mcux driver. Please see https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules for more details.

I made pr to that reppository https://github.com/zephyrproject-rtos/hal_nxp/pull/87

MaureenHelm commented:
> > Will you please add `adc` to `boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml` and check that `tests/drivers/adc/adc_api` works? The first part will cause CI to start building this driver for this board.
> 
> I have added adc to yaml file, but should I also add board to test file and adc to the board pinmux?
> 

You don't need to modify `tests/drivers/adc/adc_api/testcase.yaml`. By adding adc to the board yaml file, twister will see that the test dependency is satisfied by the board. Please do update the board pinmux.

> > You will also need to update the hal_nxp repo to build the underlying mcux driver. Please see [docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules](https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules) for more details.
> 
> I made pr to that reppository [zephyrproject-rtos/hal_nxp#87](https://github.com/zephyrproject-rtos/hal_nxp/pull/87)

Thanks

Please also squash the fixup commit into `driver: add adc driver support fot NXP RT series`. See https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow.

kele6ra commented:
> You don't need to modify `tests/drivers/adc/adc_api/testcase.yaml`. By adding adc to the board yaml file, twister will see that the test dependency is satisfied by the board. Please do update the board pinmux.
> 

I meant `tests/drivers/adc/adc_api/teast_adc.c` file, sorry for unclear message.

MaureenHelm commented on drivers/adc/Kconfig.mcux:
```diff
@@ -15,6 +15,12 @@ config ADC_MCUX_ADC16
 	help
 	  Enable the MCUX ADC16 driver.
 
+config ADC_MCUX_RTADC
```
The convention for MCUXpresso shim drivers is: `<driver class>_MCUX_<HW IP name>`. So for this driver, it should be `ADC_MCUX_12B1MSPS_SAR`. See https://github.com/zephyrproject-rtos/hal_nxp/blob/4560b0a0106bbe47eded4d7af775998c4f6d05c0/mcux/drivers/imx/fsl_adc.c#L11-L14

kele6ra replied:
I saw 3 mcux adc drivers for another nxp controllers: ADC_MCUX_ADC12, ADC_MCUX_ADC16, ADC_MCUX_LTADC. Should I rename RT driver or leave naming like other nxp mcux drivers?

MaureenHelm replied:
Yes, please rename the RT driver.

kele6ra replied:
I dhange name to driver config. If I need to rename all entries, write plese.

MaureenHelm replied:
Yes, please

kele6ra replied:
Sorry, but I can't understand what should I rename? Should I rename driver file with this name or not? Because I see that zephyr files and driver condigs don't have names with 3 underlines

MaureenHelm replied:
Yes please rename this driver with this name, the additional underscore is ok. I admit that `12B1MSPS_SAR` is a little awkward, but I want to try to reflect the name of the HW IP block because sometimes 1) the same HW IP is used in multiple SoC families (e.g., LPUART in Kinetis and i.MX RT) or 2) there are multiple HW IP block options for the same driver class (e.g., ADC12 and ADC16 in Kinetis). Hence I don't want to use the SoC family name.

kele6ra replied:
Change driver name with 12B1MSPS_SAR label, remove SoC connection in comments.

kele6ra commented on boards/arm/mimxrt1050_evk/pinmux.c:
```diff
@@ -10,11 +10,10 @@
-	.outputLogic = 0,
-	.interruptMode = kGPIO_NoIntmode
-};
+static gpio_pin_config_t enet_gpio_config = { .direction = kGPIO_DigitalOutput,
```
update file in force commit

kele6ra commented:
Hello, test fails cause of macros. These functions are used in init macros.

mmahadevan108 commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml:
```diff
@@ -26,3 +26,4 @@ supported:
   - usb_device
   - kscan:touch
   - watchdog
+  - adc
```
Please update the documentation for these boards as well including ADC in supported feature list and the pins used.

kele6ra replied:
add adc description to board doc

kele6ra commented:
Driver can't pass tests without this pr https://github.com/zephyrproject-rtos/hal_nxp/pull/87

kele6ra commented:
@MaureenHelm Hello, can you help me with https://github.com/zephyrproject-rtos/hal_nxp/pull/87 , how can I add it to actions with current adc driver?

dleach02 commented:
@kele6ra, This PR has a corresponding PR in the NXP HAL. You would typically see the west.yml changed here to reference that PR. 

https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules

where the NXP HAL section would have:

`revision: pull/18/head
`

dleach02 commented:
@MaureenHelm can you check to see if your requested changes have been resolved?

kele6ra commented:
> 
> 
> @MaureenHelm can you check to see if your requested changes have been resolved?

Thank you for help, I thought that west.yml file should be untouchable for pr for simple contributer =)

mmahadevan108 commented on west.yml:
```diff
@@ -88,7 +88,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 78efc4ba7c1057c1cf2bf06e3e27ed7cc33e1da7
+      revision: pull/87/head
```
Please separate this change to west.yml into a separate commit.

dleach02 replied:
You will need to do this because the flow is to: 

1) have this PR point to the PR in the HAL. 
2) Have the HAL PR approved and then merged.
3) Update this PR to now point a a revision SHA

kele6ra replied:
Make them separate https://github.com/zephyrproject-rtos/zephyr/pull/34078/commits/75bb9bc8fb925b08c13c5f3f9c450d65a3846626

dleach02 commented on drivers/adc/adc_mcux_12b1msps_sar.c:
```diff
@@ -0,0 +1,308 @@
+		irq_enable(DT_INST_IRQN(n));				       \
+	}
+
+DT_INST_FOREACH_STATUS_OKAY(ACD_MCUX_12B1MSPS_SAR_INIT)
```
trim new lines from the end of file

kele6ra replied:
Update in the last force commit

henrikbrixandersen commented:
I'd like to see the main commit even further (add driver, enable driver on board, enable driver in test). They are separate changes and should be handled as such.

kele6ra commented:
>
>
> I'd like to see the main commit even further (add driver, enable driver on board, enable driver in test). They are separate changes and should be handled as such.

Should I change commit comment or split commit for several commits?

henrikbrixandersen commented on drivers/adc/adc_mcux_12b1msps_sar.c:
```diff
@@ -0,0 +1,308 @@
+	if (kStatus_Success == ADC_DoAutoCalibration(base)) {
+		LOG_DBG("ADC_DoAutoCalibration() Done.");
+	} else {
+		LOG_DBG("ADC_DoAutoCalibration() Failed.");
```
```suggestion
		LOG_WRN("ADC_DoAutoCalibration() Failed.");
```

kele6ra replied:
hello, sorry for the long response, I add it to the last force update. I don't have enough time now to separate initial commit.

henrikbrixandersen reviewed (changes requested):
Apart from the logging nit and my previous remark about splitting the commit, this looks really good. Thanks!

mmahadevan108 commented:
Please add some details in the commit for the west update related to the change made in the HAL rather than the PR ID

mmahadevan108 commented:
Could you take a look at the documentation failure highlighted by the CI build

henrikbrixandersen reviewed (changes requested):
The commit log currently says:

```
fix driver copyright, fix series config,
fix dts adc address
add adc to 1050_evk board yaml
add adc params to tst_adc.c file
add 0 and 1 inputs of adc1 to pinmux file
```

Please rewrite the commit log with details of what was added and split the commit into several (at least separate driver and board changes). If in doubt, take a look at how other driver additions were handled in the Zephyr commit log.

Commit:
drivers: add adc driver for the NXP RT series

adc driver is based on the NXP MCUX hal library

Signed-off-by: Andrei Auchynnikau <ovchinnikov@strim-tech.com>

Commit:
soc: add adc driver to the nxp rt series soc

add adc support to the soc dts,
add MCUX_12B1MSPS_SAR to the soc Kconfigs

Signed-off-by: Andrei Auchynnikau <ovchinnikov@strim-tech.com>

Commit:
boards: add adc support to the mimxrt1050 board

add adc to board dts file,
update yaml files and documentation,
add adc 1 channel pins 1 and 15 to the pinmux file,
add overlay file to the board adc sample,
add board to the adc test file and

Signed-off-by: Andrei Auchynnikau <ovchinnikov@strim-tech.com>

mbolivar-nordic reviewed (approved):
DT / driver model bits lgtm

dleach02 commented:
@kele6ra, will you be able to rebase this PR to clear the issues so we can merge it?
